contracts/craft/Craft.sol

```diff
@@ -32,67 +32,81 @@ contract Craft is ICraft, CraftBase {
     function addRecipeBatch(LibCraft.Recipe[] memory _recipes) external override whenPaused() checkPermissions(MANAGER_ROLE) {
-        require(_recipes.length > 0, "Invalid input length.");
+        require(_recipes.length > 0, "Invalid input length");
 
         uint256[] memory ids = new uint256[](_recipes.length);
 
+        // Note: This allows for materials and rewards from different contracts, but this contract must be registered
+        // as a system contract on the content contract
         for (uint256 i = 0; i < _recipes.length; ++i) {
             require(_recipes[i].materials.length > 0 && _recipes[i].materials.length == _recipes[i].materialAmounts.length, "Invalid materials length");
             require(_recipes[i].rewards.length > 0 && _recipes[i].rewards.length == _recipes[i].rewardAmounts.length, "Invalid rewards length");
-            require(_recipes[i].craftingRate > 0 && _recipes[i].craftingRate <= 1000000, "Invalid crafting rate.");
+            require(_recipes[i].craftingRate > 0 && _recipes[i].craftingRate <= 1e6, "Error: Invalid crafting rate");
 
-            LibCraft.Recipe storage recipeData = recipes[recipesLength];
-            ids[i] = recipesLength;
+            LibCraft.Recipe storage recipeData = recipes[recipeCounter];
+            ids[i] = recipeCounter;
             recipeData.enabled = _recipes[i].enabled;
             recipeData.craftingRate = _recipes[i].craftingRate;
 
             for (uint256 j = 0; j < _recipes[i].materials.length; ++j) {
-                // Todo: Should we assume that these are content/ERC1155 contracts by default? Can we move
-                //       this check to the UI side?
-                require(_recipes[i].materials[j].content.supportsInterface(type(IContent).interfaceId), "Contract is not a Content Contract");
+                require(_recipes[i].materials[j].content.supportsInterface(type(IContent).interfaceId), "Error: Invalid materials contract interface");
+                require(IContent(_recipes[i].materials[j].content).isSystemContract(address(this)), "Error: Craft not registered");
 
                 recipeData.materials.push(_recipes[i].materials[j]);
                 recipeData.materialAmounts.push(_recipes[i].materialAmounts[j]);
             }
 
             for (uint256 j = 0; j < _recipes[i].rewards.length; ++j) {
-                // Todo: Should we assume that these are content/ERC1155 contracts by default? Can we move
-                //       this check to the UI side?
-                require(_recipes[i].rewards[j].content.supportsInterface(type(IContent).interfaceId), "Contract is not a Content Contract");
+                require(_recipes[i].rewards[j].content.supportsInterface(type(IContent).interfaceId), "Error: Invalid reward contract interface");
+                require(IContent(_recipes[i].rewards[j].content).isSystemContract(address(this)), "Error: Craft not registered");
 
                 recipeData.rewards.push(_recipes[i].rewards[j]);
```

Comment:
also here if rewardAmounts == 0. Unless you can do that kind of stuff in the front end instead.

Reply:
Same as the other comment. There's already a rewards.length() check above so this will never be 0.